src/repmgr_to_zk/zk.clj

```diff
@@ -1,31 +1,82 @@
 (ns repmgr-to-zk.zk
   (:require [clojure.string :as s]
+            [clojure.tools.logging :as log]
             [repmgr-to-zk.config :as config]
             [zookeeper :as zk]
-            [zookeeper.data :as zk-data]))
+            [zookeeper.data :as zk-data])
+  (:import (org.apache.zookeeper KeeperException$BadVersionException
+                                 KeeperException$SessionExpiredException
+                                 KeeperException$ConnectionLossException)))
 
-(defn close-client [client]
-  (zk/close client))
+(defonce ^{:private true} client nil)
 
-(defn get-client []
-  (zk/connect (config/lookup :zookeeper :connect)))
+(defn- deserializer [data]
+  (-> data
+      zk-data/to-string
+      read-string))
 
-(defn get-data [client path]
-  (zk/data client path))
+(defn- serializer [data]
+  (-> data
+      pr-str
+      zk-data/to-bytes))
 
-(defn create [client path]
-  (let [paths (->> (s/split path #"/")
-                   (reductions #(str %1 "/" %2))
-                   (remove s/blank?))]
-    (doseq [p paths]
-      (zk/create client p :persistent? true))))
+(def ^:const max-tries (or 3 (config/lookup :max-tries)))
 
-(defn set-master [client master-ip]
+(defn init! []
+  (alter-var-root #'client
+                  (constantly (zk/connect (config/lookup :zookeeper :connect))))
+  (log/info "Initialized zookeeper client!"))
+
+(defn destroy! []
+  (zk/close client)
+  (alter-var-root #'client nil)
+  (log/info "Destroyed zookeeper client!"))
+
+(defn- retry
+  [fn & args]
+  (loop [count max-tries]
+    (if (> count 0)
+      (let [result (try (apply fn client args)
+                        (catch KeeperException$SessionExpiredException see ::reinit)
+                        (catch KeeperException$ConnectionLossException cle ::reinit))]
+        (condp = result
+          ::reinit (do (init!)
```

Comment:
Try to log the exception. This is written in a way where we only return `::reinit` from the `catch` block and we don't have access to `see` or `cle`. We can either rewrite it to do the work right in the catch block or return a map of the result and then log the exception from it in the `condp` block.